Tokenizer almost finished

FearsomeMango73 opened the pull request:
I discovered this did not go through to Himani's Repo... trying again

This is the partially completed Tokenizer Input portion of the first set of outlines. There are two sections of suedo code that still need to be replaced in the next() of the Tokienzer.java class and the constructor of the same file I am having a little trouble with. Everything else should be there :) and I am working to figure out these issues as we speak :)
Learning more about Github everyday :P
*I was think maybe Kevin and I can talk things through on this? or if anyone wants to stay after the meeting to talk through code?*

chriscraws reviewed (changes requested):
We'll talk about best practices for Git in the meeting - this pull request currently adds 5 new files to the root directory - but these files mostly already exist in the project. Instead of creating new ones, the existing ones should be modified. That way the pull request is specific to the per-line changes you've made.

After the meeting, you should keep this code on the side and re-follow the tutorial after setting up from git closely following the guides - and as a general rule if something seems like it's going to take more than 15-30 minutes to figure out then don't hesitate to ask the group or ask me!

FearsomeMango73 committed:
Another attempt, Tokienzer, A different method hopefully works

I am uploading the two files that were added with Tokenizing Input. I
still need to update the few command lines, I just didnt want to clash
in case the files were overwritten, so I am trying this out first. A
friend recommended me to try this method... Third method is the charm!
Fingers crossed, I hope this goes through

FearsomeMango73 committed:
The Rest of Tokenizer

Here are the other changes... I dont know if it added the files, or
changed it, but I hope I did it right! Let me know! I am so happy that
is finally was working!!

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,79 @@
+  }
+  
+  public String next() throws IOException {
+  // Skip all leading whitespace
```
This comment should go below the initialization of `tokenCreation`

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,79 @@
+  
+  public String next() throws IOException {
+  // Skip all leading whitespace
+    String tokenCreation = "";
```
You can keep this initialization simple as `String tokenCreation;` since we never make use of any value before assigning it to something new

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,79 @@
+    if (remaining() <= 0) {
+      return null;
+    } else if (peek() == '"') {
+      tokenCreation += readWithNoQuotes(); //need to check
```
1. You should use `=` assignment instead of `+=` concatenation here
2. If `peek()` is returning that the next character is a `"`, then you should use your function that tokenizes input that begins with `"`. Use `readWithQuotes()` here instead :)

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,79 @@
+    } else if (peek() == '"') {
+      tokenCreation += readWithNoQuotes(); //need to check
+    } else {
+      tokenCreation += readWithQuotes(); //need to check
```
You'll need to change this line too :)

chriscraws commented on src/codeu/chat/util/TokenizerTest.java:
```diff
@@ -0,0 +1,27 @@
+  public void testWithQuotes() throws IOException {
+    final Tokenizer tokenizer = new Tokenizer(�\�hello world\� \�how are you\��);
+    assertEquals(tokenizer.next(), �hello world�);
+   assertEquals(tokenizer.next(), �how are you�);
```
Add one space here so the indentation matches up

chriscraws commented on src/codeu/chat/util/TokenizerTest.java:
```diff
@@ -0,0 +1,27 @@
+    assertEquals(tokenizer.next(), null);
+  }
+
+}
```
Add one more newline at the end, just to keep this file consistent with the others

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,79 @@
+  
+  private int remaining() {
+      return source.length() - at;
+    }
```
These two lines need their indentation fixed, also there should be an empty line after this function definition

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,79 @@
+    } else {
+      throw new IOException("End Of String");
+    }
+  }
```
One more newline here too, just for consistency!

chriscraws reviewed (changes requested):
Woo! Things look pretty good. Here's some next steps:

5 of the files are in the wrong place and/or are duplicates. These are Chat.java, Panel.java, TestRunner.java, Tokenizer.java, and TokenizerTest.java. All of them are in the base-directory of your repository. 

Tokenizer and TokenizerTest are duplicate files that can be removed:

Make sure you're in the root-directory of the repository and run `git rm TokenizerTest.java Tokenizer.java`

Chat, Panel, and TestRunner already exist in the codebase and your changes should be made to those files instead. Using copy/paste to place them in their proper locations, overwriting the old values should work. 

*After* that is done, you can run 

`git rm Chat.java Panel.java TestRunner.java`

You can commit these changes with something like `git commit -m "Moved files back into src directory"`

Once all that is done, I have a few comments on the remaining files for changes to make before you push.

Let me know if you run into any trouble or have clarifying questions!

FearsomeMango73 committed:
Deleting the Duplicates (fingers crossed this is correct :)

FearsomeMango73 committed:
Moved the rest of files back into the proper directories. I believe this is all up to date, now to fix comments!

FearsomeMango73 committed:
Fixing the Finals Changes on Tokenizer

I added the comments from Chris' request. Let me know if there is
anything else!

FearsomeMango73 committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Hayley's-Branch

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,77 @@
+//there are two parts in the next() funtion that I still need to complete
```
The next function is all done now right? you can remove this comment

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,77 @@
+  public String next() throws IOException {
+  
+    String tokenCreation;
+ // Skip all leading whitespace
```
add an indent here

chriscraws reviewed (changes requested):
Is this all wrapped up? If so, just a couple small changes and then let me know so I can run it and give LGTM :)

FearsomeMango73 committed:
Fixed final changes

chriscraws reviewed (changes requested):
All of the quotation marks are the `“` character instead of the `"` character, which is causing build-issues for me. When I replace them, I still have a list of build errors with various symbols not being found - does this build on your computer?

FearsomeMango73 committed:
Discovered true issue about the " fixed them all

This hopefully should be the final push, but I just realized that the "
were pasted in wrong. i fixed them and it should now be okay!

FearsomeMango73 committed:
Fixed bugs

There are still some errors that I can tell. From what I am seeing, The
Tokenizer was implented in Chat.java . When we made new commands with
the @overrride command, it used the old system of parsing the words. I
need to figure out how to change this: final String message =
args.hasNext() ? args.nextLine().trim() : ""; to what it needs to be?
This looks like the final step, but i got it to clean and build (with
errors, but soon to be fixed, or fixed tomorrow)

FearsomeMango73 committed:
Fixed the bugs!

With Himani's help, we were able to debug!! I have one warning (but I
think it is jsut on my computer so check) but this branch is ready to
commit and Merge! :D

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,71 @@
+  
+  private StringBuilder token;
+  private String source;
+  private int at;
```
I'd rename this variable to something a little more clear, like characterPosition

chriscraws commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,71 @@
+      if (remaining() <= 0) {
+        return null;
+     } else if (peek() == '"') {
+        tokenCreation = readWithQuotes(); //need to check
```
Not sure what the 'need to check' comment means - can it be removed?

FearsomeMango73 replied:
It has been removed, that was for my personal thought earlier on

chriscraws reviewed (changes requested):
Looks great! Just two changes then go ahead and merge it in :)

FearsomeMango73 committed:
Merge branch 'master' into Hayley's-Branch
